Add AD7091R-8 driver

machschmitt opened the pull request:
This PR does the following:
- Add a MAINTAINERS entry for AD7091R
- Add support for AD7091R-8 (experimental, not tested in hardware)

nunojsa commented on drivers/iio/adc/ad7091r-base.c:
```diff
@@ -260,7 +260,7 @@ int ad7091r_probe(struct device *dev, const char *name,
 	return devm_iio_device_register(dev, iio_dev);
 }
-EXPORT_SYMBOL_GPL(ad7091r_probe);
+EXPORT_SYMBOL_NS_GPL(ad7091r_probe, IIO_AD7091R);
```
I would drop this patch... I'm not really keen in abusing backports. Bug Fixes or important features that we really needed are of course ok to backport. This one, while nice, I'm not seeing why we can't naturally get it while moving to a newer kernel

machschmitt replied:
I must have picked that up when experimenting with newer kernels.
Sure, let's let this get naturally added to master when master be updated to a newer version.

nunojsa commented on MAINTAINERS:
```diff
@@ -1103,6 +1103,16 @@ S:	Supported
 F:	drivers/iio/dac/ad5770r.c
 
+ANALOG DEVICES INC AD7091R DRIVER
+M:	Beniamin Bia <beniamin.bia@analog.com>
```
Beniamin Bia is no longer in ADI... I guess you can put yourself in here :)

machschmitt replied:
First driver inherited, yay :)

nunojsa commented on drivers/iio/adc/ad7091r8.c:
```diff
@@ -0,0 +1,200 @@
+const struct regmap_access_table ad7091r4_readable_regs_table = {
+	.yes_ranges = ad7091r4_readable_ranges,
+	.n_yes_ranges = ARRAY_SIZE(ad7091r4_readable_ranges),
+};
```
ditto...

nunojsa commented on drivers/iio/adc/ad7091r8.c:
```diff
@@ -0,0 +1,200 @@
+
+	chip_info = device_get_match_data(&spi->dev);
+	if (!chip_info)
+		chip_info = (const struct ad7091r_chip_info *)spi_get_device_id(spi)->driver_data;
```
sanity check again `chip_info`

nunojsa commented on drivers/iio/adc/ad7091r8.c:
```diff
@@ -0,0 +1,200 @@
+	if (IS_ERR(map)) {
+		dev_err(&spi->dev, "Error initializing spi regmap: %ld\n",
+			PTR_ERR(map));
+		return PTR_ERR(map);
```
`dev_err_probe()`

nunojsa commented on arch/arm/boot/dts/overlays/rpi-ad7091r8-overlay.dts:
```diff
@@ -0,0 +1,24 @@
+			};
+		};
+	};
+};
```
you're right in your assumption... This patch should go to the rpi branch

machschmitt replied:
Okay, so ... how does one submit a raspberrypi overlay to ADI repo?
Do I just create a PR to rpi-5.15.y only containing the overlay?

nunojsa replied:
yes...

nunojsa commented on drivers/iio/adc/ad7091r-base.c:
```diff
@@ -59,10 +59,14 @@ static int ad7091r_set_mode(struct ad7091r_state *st, enum ad7091r_mode mode)
-		return ret;
+	/* AD7091R-2/4/8 don't set normal, command, autocycle modes in conf reg */
+	if (st->chip_info->type == AD7091R5) {
+		return 0;
```
?

machschmitt replied:
Yes, that's where I kind of got stuck.
AD7091R-5 and AD7091R-2/-4/-8 are very similar except for the layout of the configuration register and the operating modes they support. AD7091R-5 has four operating modes (sample, command, autocycle, power-down) while R-8 has two (normal (equivalent to R-5 sample mode), and power-down). The config modes are set by writing to the config register and bits related to modes that are unavailable to R-8 devices are labeled as reserved in R-8 datasheet. Despite that, ad7091r-base could stay pretty much unchanged except for some minor tweaks. That came at the cost of adding checks for those few divergences between designs.

One alternative would be defining a set of reg_field and replacing regmap_read/write with regmap_field_read/write when appropriate. I'm drawing inspiration from fxas21002c_core.c, except that R-5 and R-8 would each have their own set of reg_field declarations. Problem is, for the config register, we don't want to read or write from/to fields that are differently laid out within a register. It's a read/write or don't read/write case. Also thought of defining an ops struct (like sps30), but that would probably be overkill.

Moreover, it looks like the default regmap_write() implementation for the SPI bus does a spi_write() with reg address set as tx and then does another spi_write() with the value set in tx. Though, AD7091R-8 protocol (datasheet page 31) is to do a single spi transfer with the address set in bits B15:B11 and value set in bits B9:B0. Similar issue with default regmap_read(). So, I wonder if R-8 won't need a custom regmap_read/write implementation by setting regmap_config.reg_read/write.

nunojsa replied:
Hmmm I see now why it's a draft... I was thinking the original driver was writing on reserved bits.

> Moreover, it looks like the default regmap_write() implementation for the SPI bus does a spi_write() with reg address set as tx and then does another spi_write() with the value set in tx.

Hmm, where did you saw that? AFAIK, only one spi_write() is done for regmap_write() and spi_write_then_read() for regmap_read(). Often the issue is on the read side for devices that need to deassert + assert again the CS pin.
  
> B15:B11 and value set in bits B9:B0. Similar issue with default regmap_read(). So, I wonder if R-8 won't need a custom regmap_read/write implementation by setting regmap_config.reg_read/write.

If needed, do a custom bus instead a complete bypass. With the bus, you can still benefit from all the regmap sutff... I did it (was advised to do it) in [ltc2688.c](https://elixir.bootlin.com/linux/v6.3-rc5/source/drivers/iio/dac/ltc2688.c).

nunojsa commented on drivers/iio/adc/ad7091r-base.c:
```diff
@@ -59,10 +59,14 @@ static int ad7091r_set_mode(struct ad7091r_state *st, enum ad7091r_mode mode)
+		ret = regmap_update_bits(st->map, AD7091R_REG_CONF,
+					 AD7091R_REG_CONF_MODE_MASK, conf);
+		if (ret)
+			return ret;
```
no one ever complained about this so I'm not sure if writing these bits is problematic (the device might just discard it)...

Anyways, the patch is obviously good... I'm just wondering if it deserves a Fixes tag or not. Maybe not, but let's see what upstream has to say

machschmitt replied:
It's correct for AD7091R-5 devices so I wouldn't give it a fixes tag.
Well if it's certain that -2/-4/-8 devices would discard bits written to those reserved spots of the config register, then that would save us some time and driver code.
I often tend to be more pessimistic so I added the check (just in case ...).
I had left this in a separate commit but am now thinking it could be squashed with the patch adding AD7091R-8. AD7091R-8 may get added with buggy behavior otherwise.

nunojsa commented on drivers/iio/adc/ad7091r-base.c:
```diff
@@ -106,8 +107,13 @@ static int ad7091r_read_one(struct iio_dev *iio_dev,
+	} else {
+		if (AD7091R8_REG_RESULT_CH_ID(val) != channel)
+			return -EIO;
+	}
```
Hmm, I would say you could squash this with the patch adding support for AD7091R-8

machschmitt replied:
Yes, same reason the other "fix".

machschmitt commented:
Hi Nuno, thanks for your accurate review.

Having the opportunity to send this series upstream is something I look forward :)
Though, one thing I'd like to do first is to get it better tested (particularly the regmap stuff).
It's a bit cloudy to me whether the current driver would be able to actually manage AD7091R-8 devices.
Dragos is seeing to ship me an evaluation board for that design.
Though, before having a device to partner with the driver, it's hard to assert reads and writes will work properly.

Anyway, thank you :)

machschmitt commented:
Change log V1 -> V2

- Dropped upstream patch that will get pulled when master branch gets updated to a newer kernel version.
- Replaced Beniamin by me in the MAINTAINERS file entry for AD7091R
- Droped vref-supply trivial comment.
- Added pattern properties to describe device channels.
- Removed example device node phandle.
- Declared regmap access tables as static members.
- Added sanity check for device chip info.
- Used dev_err_probe to do a cleaner error handling at device probe.
- Squashed two separate commits into the add driver commit
- Removed rpi overlay from the series

nunojsa commented:
> Though, one thing I'd like to do first is to get it better tested (particularly the regmap stuff). 

Oh, sure... It should be tested of course :)

machschmitt commented:
Change log V2 -> V3

- Refactored device probe functions to allocate IIO device struct before calling the generic AD7091R probe.
- Implemented a custom regmap_bus for appropriate handle of read/write operations.
- Added device tree file for AD7091R-8 with Cora Z7.

machschmitt commented:
Change log V2 -> V3

- Implemented custom regmap bus support.
- Added CONVST GPIO control for initiating ADC conversions.
- Added RESET GPIO control to pulse reset on power up (required for proper initialization).
- Added direct register access through debugfs.
- Enabled event support

Changes tested with EVAL-AD7091R-xSDZ and Raspberry Pi 4.
Wiki page: https://wiki.analog.com/resources/tools-software/linux-drivers/iio-adc/ad7091r8

machschmitt commented:
The 2 channel variant (AD7091R-2) doesn't have ALERT/BUSY/GPO pin. I want to add a dt-schema check to complain if one adds ALERT interrupt for a 2 channel ADC (something more or less like [ad7476](https://github.com/analogdevicesinc/linux/blob/master/Documentation/devicetree/bindings/iio/adc/adi%2Cad7476.yaml#L153) ). I'm struggling to achieve that though.

We may do event support in a different way if there is a way to set an interrupt for a GPIO pin. I just didn't quickly find how to do that so I went for something similar to ad7091r-5 (the I2C variant).

machschmitt commented:
Hi @nunojsa, I believe this is ready for review.

The driver has been tested with EVAL-AD7091R-xSDZ and Raspberry Pi 4.
Wiki page: https://wiki.analog.com/resources/tools-software/linux-drivers/iio-adc/ad7091r8

The only thing remaining is to get a dt-scheme/binding constraint correct.
I want to add a dt-schema check to prevent configuring the ALERT interrupt on ADCs that don't have an interrupt pin. (something more or less like [ad7476](https://github.com/analogdevicesinc/linux/blob/master/Documentation/devicetree/bindings/iio/adc/adi%2Cad7476.yaml#L153) ). Can you give me any hints on how to implement that check?

machschmitt commented:
> > Can you give me any advice on how to implement that check?
> 
> Can't you do it as in the example you pointed me? I would suggest exactly the same story with `-if` and `compatibles`

Well yes, I did that but dt_binding_check was reporting a strange warning. I thought it was because of something wrong with the `-if` but I finally realized it was because I had `$ref: /schemas/spi/spi-peripheral-props.yaml#` in the binding and that was causing a warning in kernel prior to 6.1. I rebased on top of master and now dt_binding_check is ok with that.

nunojsa commented on drivers/iio/adc/ad7091r8.c:
```diff
@@ -0,0 +1,340 @@
+{
+	struct ad7091r_state *st = context;
+	struct spi_device *spi = container_of(st->dev, struct spi_device, dev);
+	u32 write_data = *((int *)data);
```
another possible unaligned access...

nunojsa commented on drivers/iio/adc/ad7091r8.c:
```diff
@@ -0,0 +1,340 @@
+	struct spi_device *spi = container_of(st->dev, struct spi_device, dev);
+	unsigned int reg_add = *(unsigned int *)reg & 0x1f;
+	__be16 rx;
+	__be16 tx;
```
both buffers are not DMA safe. Jonathan is very strict regarding this. So, even though it's just two bytes, you need to make it DMA safe or you might get complains...

nunojsa commented on drivers/iio/adc/ad7091r8.c:
```diff
@@ -0,0 +1,340 @@
+	 */
+	tx = cpu_to_be16(FIELD_PREP(AD7091R8_REG_DATA_MSK, tx_data) |
+			 FIELD_PREP(AD7091R8_RD_WR_FLAG_MSK, 1) |
+			 FIELD_PREP(AD7091R8_REG_ADDR_MSK, reg_add));
```
this is odd... make sure you look at `regmap_config` and see if you can use it's fields properly. I'm not 100% sure but I suspect the above could really be just `spi_write(spi, data, count)`

nunojsa commented on drivers/iio/adc/ad7091r8.c:
```diff
@@ -0,0 +1,340 @@
+	if (ret < 0)
+		return ret;
+
+	ret = be16_to_cpu(rx);
```
In theory you don't need to deal with this... regmap should already take care of it. Set `REGMAP_ENDIAN_BIG` in your bus. I might be missing something but I believe you can make these write/read functions simpler by letting the core do more work.

nunojsa reviewed (commented):
Ok, some minor things... The most tricky would be the regmap_bus. I think that one still needs more thought and work.

Anyways, feel free to take my inputs, refactor it as you want and send it upstream

machschmitt commented:
Support for AD7091R-2/-4/-8 is going through review in the Linux kernel community mailing lists.
This PR is now outdated and has lost its purpose.
